client/state/data-layer/wpcom/sites/hosting/php-version.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * Internal dependencies
+ */
+import { http } from 'state/data-layer/wpcom-http/actions';
+import { dispatchRequest } from 'state/data-layer/wpcom-http/utils';
+import { registerHandlers } from 'state/data-layer/handler-registry';
+import {
+	HOSTING_PHP_VERSION_GET,
+	HOSTING_PHP_VERSION_GET_SUCCESS,
+	HOSTING_PHP_VERSION_SET,
+	HOSTING_PHP_VERSION_SET_FAILURE,
+	HOSTING_PHP_VERSION_SET_SUCCESS,
+} from 'state/action-types';
+import { errorNotice, successNotice } from 'state/notices/actions';
+import { translate } from 'i18n-calypso';
+import { composeAnalytics, recordGoogleEvent, recordTracksEvent } from 'state/analytics/actions';
+
+const updateNoticeId = 'hosting-php-version';
+
+const getPhpVersion = action =>
+	http(
+		{
+			method: 'GET',
+			path: `/sites/${ action.siteId }/hosting/php-version`,
+			apiNamespace: 'wpcom/v2',
+			body: {},
+		},
+		action
+	);
+
+const getPhpVersionSuccess = ( action, version ) => {
+	return {
+		type: HOSTING_PHP_VERSION_GET_SUCCESS,
+		siteId: action.siteId,
+		version,
+	};
+};
+
+const updatePhpVersion = action =>
+	http(
+		{
+			method: 'POST',
+			path: `/sites/${ action.siteId }/hosting/php-version`,
+			apiNamespace: 'wpcom/v2',
+			body: {
+				version: action.version,
+			},
+		},
+		action
+	);
+
+export const hostingPhpVersionUpdateTracking = ( version, result ) =>
+	composeAnalytics(
+		recordGoogleEvent(
+			'Hosting Configuration',
+			'Clicked "Update PHP Version" Button in PHP Version box',
+			`PHP Version Update ${ result }`,
+			version
+		),
+		recordTracksEvent( 'calypso_hosting_configuration_php_version_update', {
+			result,
+			version,
+		} )
+	);
+
+const updatePhpVersionSuccess = action => {
+	return [
+		hostingPhpVersionUpdateTracking( action.version, true ),
+		{
+			type: HOSTING_PHP_VERSION_SET_SUCCESS,
+			siteId: action.siteId,
+			version: action.version,
+		},
+		successNotice(
+			translate( 'PHP version successfully set to %(version)s.', {
+				args: {
+					version: action.version,
+				},
+			} ),
+			{
+				id: updateNoticeId,
+				showDismiss: false,
+				duration: 5000,
+			}
+		),
+	];
+};
+
+const updatePhpVersionError = action => {
+	return [
+		hostingPhpVersionUpdateTracking( action.version, false ),
+		errorNotice( translate( 'Failed to set PHP version.' ), {
+			id: updateNoticeId,
+		} ),
+		{
+			type: HOSTING_PHP_VERSION_SET_FAILURE,
+			siteId: action.siteId,
+		},
+	];
+};
+
+registerHandlers( 'state/data-layer/wpcom/sites/hosting/php-version.js', {
+	[ HOSTING_PHP_VERSION_SET ]: [
```

Comment:
```jsx
[ HOSTING_PHP_VERSION_SET ]: [ // Let the reducer also read this, it counts as an optimistic update. The optimistic part is optional.
    dispatchRequest( {
			fetch: updatePhpVersion,
			onSuccess: showSuccessNotice,
			onError: showErrorAndRestoreVersion,
		} ),
]
```

```jsx
const showErrorAndRestoreVersion = action => {
    return [
        bypassDataLayer( setPHPVersion( lastVersion ) ), // Restore last state on error. Since this isn't a boolean, we need to pass a little more state in the action for the current php version if the error doesn't have this. Note that bypassDataLayer adds meta to tell the datalayer to not fire the handler associated with "HOSTING_PHP_VERSION_SET" (avoids an infinite loop of HTTP calls).
		errorNotice( translate( 'Failed to set PHP version.' ), {
			id: updateNoticeId,
		} ),
    ];
}
```

Reply:
I don't like this. Bypassing for the sake of one action less, in a tbh very convoluted way for me.

Reply:
If we're going the non-optomistic route (which I think maps better to deletes, and where we want to update UI instantly), not too much changes besides a new action:

```jsx
HOSTING_PHP_VERSION_REQUEST, // only used in data-layer, http mapping
HOSTING_PHP_VERSION_SET_REQUEST, //only used in data-layer http mapping
HOSTING_PHP_VERSION_SET, // meaningful app action, read by the reducer ✨ 
```

Reply:
I like this one. :bowing_man: thank you.